File: backend/app/api/v1/courses.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app.core.supabase import get_supabase

router = APIRouter(prefix="/courses", tags=["courses"])


class CourseSummary(BaseModel):
    slug: str
    title: str
    summary: str
    level: str                       # beginner / intermediate / advanced
    duration_hours: int = 0
    price: int | None = None
    thumbnail_url: str | None = None
    sort_order: int = 0


class CourseDetail(CourseSummary):
    description: str = ""


_LIST_COLUMNS = "slug,title,summary,level,duration_hours,price,thumbnail_url,sort_order"
_DETAIL_COLUMNS = _LIST_COLUMNS + ",description"


@router.get("", response_model=list[CourseSummary])
async def list_courses() -> list[CourseSummary]:
    response = (
        get_supabase()
        .table("courses")
        .select(_LIST_COLUMNS)
        .eq("status", "published")
        .order("sort_order")
        .execute()
    )
    return [CourseSummary(**row) for row in response.data or []]


@router.get("/{slug}", response_model=CourseDetail)
async def get_course(slug: str) -> CourseDetail:
    response = (
        get_supabase()
        .table("courses")
        .select(_DETAIL_COLUMNS)
        .eq("slug", slug)
        .eq("status", "published")
        .limit(1)
        .execute()
    )
    rows = response.data or []
    if not rows:
        raise HTTPException(status_code=404, detail="course not found")
    return CourseDetail(**rows[0])

File: frontend/src/lib/courses.ts
import "server-only";

import { supabase } from "@/lib/supabase";

export type CourseLevel = "beginner" | "intermediate" | "advanced";

export type CourseSummary = {
  slug: string;
  title: string;
  summary: string;
  level: CourseLevel;
  duration_hours: number;
  price: number | null;
  thumbnail_url: string | null;
  sort_order: number;
};

export type CourseDetail = CourseSummary & {
  description: string;
};

const LIST_COLUMNS =
  "slug,title,summary,level,duration_hours,price,thumbnail_url,sort_order";
const DETAIL_COLUMNS = `${LIST_COLUMNS},description`;

export async function fetchCourses(): Promise<CourseSummary[]> {
  const { data, error } = await supabase
    .from("courses")
    .select(LIST_COLUMNS)
    .eq("status", "published")
    .order("sort_order", { ascending: true });
  if (error) throw new Error(`Failed to fetch courses: ${error.message}`);

  return (data ?? []).map((row) => ({
    slug: row.slug as string,
    title: (row.title as string) ?? "",
    summary: (row.summary as string) ?? "",
    level: row.level as CourseLevel,
    duration_hours: (row.duration_hours as number) ?? 0,
    price: (row.price as number | null) ?? null,
    thumbnail_url: (row.thumbnail_url as string | null) ?? null,
    sort_order: (row.sort_order as number) ?? 0,
  }));
}

export async function fetchCourse(slug: string): Promise<CourseDetail | null> {
  const safe = decodeURIComponent(slug);
  const { data, error } = await supabase
    .from("courses")
    .select(DETAIL_COLUMNS)
    .eq("slug", safe)
    .eq("status", "published")
    .limit(1)
    .maybeSingle();
  if (error) {
    if (error.code === "PGRST116") return null;
    throw new Error(`Failed to fetch course ${slug}: ${error.message}`);
  }
  if (!data) return null;

  return {
    slug: data.slug as string,
    title: (data.title as string) ?? "",
    summary: (data.summary as string) ?? "",
    level: data.level as CourseLevel,
    duration_hours: (data.duration_hours as number) ?? 0,
    price: (data.price as number | null) ?? null,
    thumbnail_url: (data.thumbnail_url as string | null) ?? null,
    sort_order: (data.sort_order as number) ?? 0,
    description: (data.description as string) ?? "",
  };
}
